receivequeue.go

```diff
@@ -0,0 +1,147 @@
+package multipath
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+// receiveQueue keeps received frames for the upper layer to read. It is
+// maintained as a ring buffer with fixed size. It takes advantage of the fact
+// that the frame number is consequential, so when a new frame arrives, it is
+// placed at the position indexed by the remainder of the frame number divided
+// by the buffer size.
+type receiveQueue struct {
+	buf  []frame
+	size uint64
+	// rp stands for read pointer, point to the index of the frame containing
+	// data yet to be read.
+	rp           uint64
+	cond         *sync.Cond
+	condAdd      *sync.Cond
+	readDeadline time.Time
+	closed       uint32 // 1 == true, 0 == false
+}
+
+func newReceiveQueue(size int) *receiveQueue {
+	rq := &receiveQueue{
+		buf:     make([]frame, size),
+		size:    uint64(size),
+		rp:      minFrameNumber % uint64(size), // frame number starts with minFrameNumber, so should the read pointer
+		cond:    sync.NewCond(&sync.Mutex{}),
+		condAdd: sync.NewCond(&sync.Mutex{}),
+	}
+	return rq
+}
+
+func (rq *receiveQueue) add(f *frame) {
+	for {
+		if rq.tryAdd(f) {
+			return
+		}
+		log.Error("would override unconsumed frame! wait for available slot")
```

Comment:
I thought that if it happens too often, it would degrade throughput, so the frequency of the error would signal us to adjust the buffer size. But you are right, if the caller doesn't consume data quick enough, having a larger buffer does no help but increase memory profile. Will remove.